test/unit/groovy/PromotionHelperTest.groovy

```diff
@@ -0,0 +1,74 @@
+package unit.groovy
+
+import com.lesfurets.jenkins.unit.BasePipelineTest
+import hudson.model.Hudson
+import hudson.model.TopLevelItem
+import jenkins.model.Jenkins
+import org.jenkinsci.plugins.workflow.job.WorkflowJob
+import org.jenkinsci.plugins.workflow.job.WorkflowRun
+import org.jenkinsci.plugins.workflow.support.steps.build.RunWrapper
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.powermock.api.mockito.PowerMockito
+import org.powermock.core.classloader.annotations.PrepareForTest
+import org.powermock.modules.junit4.PowerMockRunner
+
+import static org.junit.Assert.assertEquals
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest([Jenkins.class, RunWrapper.class, WorkflowJob.class, WorkflowRun.class])
+class PromotionHelperTest extends BasePipelineTest {
+
+    private Script promotionHelper
+    private RunWrapper currentBuild
+    private Hudson instance
+    private ArrayList<RunWrapper> upstreamBuilds
+
+    @Override
+    @Before
+    void setUp() throws Exception {
+        super.setUp();
+        promotionHelper = loadScript('vars/promotionHelper.groovy')
+
+        currentBuild = PowerMockito.mock(RunWrapper.class)
+        instance = PowerMockito.mock(Hudson.class)
+
+        PowerMockito.mockStatic(Jenkins.class);
+        PowerMockito.when(Jenkins.getInstanceOrNull()).thenReturn(instance)
+
+        upstreamBuilds = new ArrayList<>()
+    }
+
+    @Test
+    void getUpstreamTriggerOrLatestBuildNumber_ReturnsUpstreamBuildNumber_WithAutomaticTrigger() throws Exception {
+        // when triggered automatically, currentBuild contains an array of upstream builds, which can be inspected to
+        // retrieve the build number
+        RunWrapper upstreamBuild = PowerMockito.mock(RunWrapper.class)
+        PowerMockito.when(upstreamBuild.number).thenReturn(666)
+        PowerMockito.when(upstreamBuild.getFullDisplayName()).thenReturn('upstream_job #666')
+
+        upstreamBuilds.add(upstreamBuild)
```

Comment:
Would be good to add multiple upstream builds to test that the correct one is retrieved.

Reply:
Yep fair point, since it is explicitly only retrieving from the first build. There's only going to be one build in our use case, make makes sense to add this anyway to be specific.